File: tools/omedia.nim
#[
omedia organizes photo and video files into a directory tree by date
Usage: omedia -i:<sourceDir> -o:<sinkDir>
Dependencies
- exiv2 extracts date and time from a photo file
- ffprobe (from ffmpeg) extracts date and time from a video file
]#

import std/[strformat, strutils, sequtils]
import std/[parseopt, threadpool, os, osproc, times, sha1]
import pkg/regex

type
  MediaType = enum mtImage, mtVideo
  Media = object
    mType: MediaType
    source: string
    tStamp: DateTime
    digest: string
  OptionsError = object of ValueError
  FSError = object of OSError

const
  reImageSource = re"\.(?i:jpe?g)"
  cmdImageTStamp = "exiv2 -K Exif.Image.DateTime -P v"
  reImageTStamp = re"(?P<tStamp>\d{4}:\d{2}:\d{2} \d{2}:\d{2}:\d{2})"
  fmtImageTStamp = "yyyy:MM:dd HH:mm:ss"
  reVideoSource = re"\.(?i:mp4|mov|avi)"
  cmdVideoTStamp = "ffprobe -v quiet -select_streams v:0 -show_entries stream_tags=creation_time -of default=nw=1:nk=1"
  reVideoTStamp = re"(?P<tStamp>\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2})"
  fmtVideoTStamp = "yyyy-MM-dd'T'HH:mm:ss"

proc parseOptions(): tuple[input, output: string] =
  var input, output: string
  for _, key, value in getOpt():
    case key:
    of "i", "input": input = value
    of "o", "output": output = value
  if input == "" or output == "":
    raise newException(OptionsError, "options -i and -o are mandatory")
  if not dirExists(input):
    raise newException(FSError, "input directory does not exist: " & input)
  createDir(output)
  (input, output)

iterator walkMedia(sourceDir: string,
                   reImage = reImageSource,
                   reVideo = reVideoSource): Media =
  for file in walkDirRec(sourceDir):
    if file.endsWith(reImage):
      yield Media(mType: mtImage, source: file)
    elif file.endsWith(reVideo):
      yield Media(mType: mtVideo, source: file)
    else: stderr.write(fmt "WARNING: unknown media type: {file}\n")

proc extractImageTStamp(media: var Media,
                        lastMediaTStamp: var DateTime,
                        log: var seq[string],
                        cmdTStamp = cmdImageTStamp,
                        reTStamp = reImageTStamp,
                        fmtTStamp = fmtImageTStamp) =
  let (output, exitCode) =
    execCmdEx(fmt "{cmdTStamp} {quoteShell media.source}")
  if exitCode != 0:
    log.add(fmt "WARNING: exiv2: non-zero exit code: {exitCode}")
  var m: RegexMatch
  if output.find(reTStamp, m):
    media.tStamp = m.groupFirstCapture("tStamp", output).parse(fmtTStamp)
    lastMediaTStamp = media.tStamp
  else:
    media.tStamp = lastMediaTStamp

proc extractVideoTStamp(media: var Media,
                        lastMediaTStamp: var DateTime,
                        log: var seq[string],
                        cmdTStamp = cmdVideoTStamp,
                        reTStamp = reVideoTStamp,
                        fmtTStamp = fmtVideoTStamp) =
  let (output, exitCode) =
    execCmdEx(fmt "{cmdTStamp} {quoteShell media.source}")
  if exitCode != 0:
    log.add(fmt "WARNING: ffprobe: non-zero exit code: {exitCode}")
  var m: RegexMatch
  if output.find(reTStamp, m):
    media.tStamp = m.groupFirstCapture("tStamp", output).parse(fmtTStamp)
    lastMediaTStamp = media.tStamp
  else:
    media.tStamp = lastMediaTStamp

proc digestMedia(media: var Media) =
  media.digest = media.source.secureHashFile.`$`.toLowerAscii

proc writeMedia(media: Media, sinkDir: string) =
  let mediaDir = media.tStamp.format("yyyy/yyyy-MM-dd")
  let mediaExt = media.source.splitFile[2].toLowerAscii.replace(re"jpeg$", "jpg")
  let mediaFile = media.tStamp.format("yyyyMMdd'_'HHmmss'_'") &
    media.digest & mediaExt
  createDir(sinkDir / mediaDir)
  copyFile(media.source, sinkDir / mediaDir / mediaFile)

proc organizeMedia(mediaGroup: seq[Media], sinkDir: string) =
  var lastMediaTStamp = now()
  for media in mediaGroup:
    var
      media = media
      log = newSeq[string](3)
    try:
      case media.mType:
      of mtImage:
        log.add(fmt "IMAGE: {media.source}")
        extractImageTStamp(media, lastMediaTStamp, log)
      of mtVideo:
        log.add(fmt "VIDEO: {media.source}")
        extractVideoTStamp(media, lastMediaTStamp, log)
      digestMedia(media)
      writeMedia(media, sinkDir)
    except CatchableError as error:
      log.add(fmt "ERROR: {error.msg}")
    echo log.join("\n").strip

try:
  let
    (inputDir, outputDir) = parseOptions()
    cpuCores = countProcessors()
    mediaGroups = inputDir.walkMedia.toSeq.distribute(cpuCores)
  for mediaGroup in mediaGroups:
    spawn organizeMedia(mediaGroup, outputDir)
  sync()
except OptionsError as error:
  stderr.write fmt "ERROR: {error.msg}\n"
  quit("Usage: omedia -i:<sourceDir> -o:<sinkDir>")

File: tools/omedia.js
#!/usr/bin/env node

/*
 * omedia organizes photo and video files into a directory tree by date
 * Usage: omedia -i <sourceDir> -o <sinkDir> -j <jobs>
 * Dependencies
 * - exiv2 extracts date and time from a photo file
 * - ffprobe (from ffmpeg) extracts date and time from a video file
 * - openssl generates a digest of a media file
 * cp -r ~/Media/sink/2022/* / /run/media/vlad/SD2/Media-2021-202_/2022
 */

import { DateTime } from "luxon"
import { extname } from "path"
import { copyFile } from "fs/promises"
import { pathExists, mkdirp } from "fs-extra"
import { globbyStream } from "globby"
import chalk from "chalk"
import parseArgs from "minimist"
import { $ } from "zx"

$.verbose = false
let lastTs = DateTime.now()

function parseTs(str, format) {
  const ts = DateTime.fromFormat(str, format)
  if (ts.invalid) {
    throw new Error(ts.invalid.explanation)
  }
  return ts
}

async function fromReadable(readable) {
  let res = ""
  for await (const chunk of readable) { res += chunk }
  return res
}

function asyncParallel(jobs, limit) {
  let index = 0
  let running = 0
  let completed = 0
  return new Promise((resolve, reject) => {
    function done(error) {
      if (error) { console.error(chalk.red(error)) }
      if (++completed === jobs.length) { return resolve() }
      if (--running < limit) { parallel() }
    }
    function parallel() {
      while (index < jobs.length && running < limit) {
        const job = jobs[index++]
        job().then(done, done)
        ++running
      }
    }
    parallel()
  })
}

async function extractImageTs(file, logs) {
  try {
    const res = await $`exiv2 -K Exif.Image.DateTime -P v ${file}`
    let output = await fromReadable(res.stdout)
    output = output.replace("\n", "")
    return parseTs(output, "yyyy:MM:dd HH:mm:ss")
  } catch (error) { logs.push(chalk.red(error)) }
}

async function extractVideoTs(file, logs) {
  try {
    const res = await $`ffprobe -v quiet -select_streams v:0 -show_entries stream_tags=creation_time -of default=nw=1:nk=1 ${file}`
    let output = await fromReadable(res.stdout)
    output = output.replace(/\.\d+Z$\n/m, "")
    return parseTs(output, "yyyy-MM-dd'T'HH:mm:ss")
  } catch(error) { logs.push(chalk.red(error)) }
}

async function digestMedia(file) {
  const res = await $`openssl dgst -sha256 ${file}`
  const output = await fromReadable(res.stdout)
  return output.replace(/^.+= /, "").slice(0, 16)
}

async function writeMedia(file, ts, digest, args) {
  const dir = ts.toFormat("yyyy/yyyy-MM-dd")
  const fileTime = ts.toFormat("yyyyMMdd_HHmmss")
  const ext = extname(file).toLowerCase().replace("jpeg", "jpg")
  await mkdirp(`${args.o}/${dir}`)
  await copyFile(file, `${args.o}/${dir}/${fileTime}_${digest}${ext}`)
}

async function organizeMedia(file, type, extractTs, args) {
  const logs = []
  try {
    logs.push(`${chalk.green(type)} ${file}`)
    let ts = await extractTs(file, logs)
    if (ts) { lastTs = ts } else { ts = lastTs }
    const digest = await digestMedia(file)
    await writeMedia(file, ts, digest, args)
  } catch (error) { logs.push(chalk.red(error)) }
  console.log(logs.join("\n"))
}

async function readMedia(args) {
  const jobs = []
  if (!await pathExists(args.i)) {
    throw new Error(`Input directory ${args.i} does not exist`)
  }
  for await (const file of globbyStream(args.i)) {
    if (/(jpe?g|heic)$/i.test(file)) {
      jobs.push(() => organizeMedia(file, "image", extractImageTs, args))
    } else if (/(mp4|mov|avi)$/i.test(file)) {
      jobs.push(() => organizeMedia(file, "video", extractVideoTs, args))
    } else {
      console.log(chalk.yellow("unknown"), file)
    }
  }
  return jobs
}

async function organize(args) {
  const jobs = await readMedia(args)
  return await asyncParallel(jobs, args.j)
}

function configure() {
  const argsConfig = {
    alias: { i: "input", o: "output", j: "jobs" },
    default: { i: ".", o: "sink", j: 128 }
  }
  return parseArgs(process.argv.slice(2), argsConfig)
}

await organize(configure())
